src/hw/hw_mngr.cpp

```diff
@@ -357,3 +509,177 @@ void HWManager::setFanSpeedDelayed()
         setFanSpeed();
     });
 }
+
+void HWManager::detectFansDelayed(uint32_t delaySecs)
+{
+    static boost::asio::deadline_timer tmr(io);
+
+    //  if we has expired attempts, then stop fan detection 
+    //  and work in normal state
+    if (++numErrorAttempts > maxErrorAttempts)
+    {
+        log<level::ERR>("Reached maxErrorAttempts",
+                        entry("STATE=%d", int(detectFansState)));
+        detectFansState = FanState::normal;
+        return;
+    }
+
+    // this implicitly cancels the timer
+    tmr.expires_from_now(boost::posix_time::seconds(delaySecs));
+    tmr.async_wait([&](const boost::system::error_code& err) {
+        if (err == boost::asio::error::operation_aborted)
+        {
+            /* we were canceled*/
+            return;
+        }
+        else if (err)
+        {
+            log<level::ERR>("Timer error",
+                            entry("WHAT=%s", err.message().c_str()));
+            return;
+        }
+
+        processDetectState();
+    });
+}
+
+void HWManager::onHostPowerOn()
+{
+    log<level::DEBUG>("Host power is ON");
+    detectFansState = FanState::uninit;
+    numErrorAttempts = 0;
+}
+
+void HWManager::onHostPowerOff()
+{
+    log<level::DEBUG>("Host power is OFF");
+}
+
+void HWManager::runDetectFans()
+{
+    if (detectFansState == FanState::uninit)
+    {
+        detectFansState = FanState::init;
+        processDetectState();
+    }
+}
+
+void HWManager::processDetectState()
+{
+    if (detectFansState == FanState::normal)
+        return;
+ 
+    //  wait until createInventory() complete a work
+    if (!config.desc)
+    {
+        detectFansDelayed(5);
+        return;
+    }
+
+    if (!processSystemFans())
+    {
+        detectFansState = FanState::normal;
+        return;
+    }
+
+    switch (detectFansState)
+    {
+    case FanState::init:
+        detectFansState = FanState::detect;
+        detectFansDelayed(5);
+        break;
+    case FanState::detect:
+        updateFansInfo();
+        detectFansState = FanState::normal;
+        break;
+    default:
+        break;
+    }
+}
+
+bool HWManager::processSystemFans()
+{
+    std::string  path = getFansControlPath();
+    if (path.empty())
+        return false;
+
+    std::stringstream ssLog;
+
+    for (const auto& [conIndex, conDescr] : config.desc->fans)
+    {
+        if (conDescr.type != ConnectorType::SYSTEM)
+            continue;
+
+        FanFeature& fan = fanFeatures[conIndex];
+        auto pwmPath = path + "pwm" + std::to_string(conDescr.fanIndex);
+        auto inletPath = path + "fan" + std::to_string(conDescr.tachIndexA+1) + "_input";
+        auto outletPath = path + "fan" + std::to_string(conDescr.tachIndexB+1) + "_input";
+
+        switch (detectFansState)
+        {
+        case FanState::init:
+            fan.initialPwm = getHwmonValue(pwmPath);
+            fan.maxInletRpm = getHwmonValue(inletPath);
```

Comment:
why do you save RPMs here? They are not "MAX" on this stage, isn't it? Well, you can check if "initilPWM" already at 100% and then skip next stage, but in current implementation it looks useless.

Reply:
there is algorithm which is described in the task 1141